### ISSUE TITLE
[Form] Placeholder handling broken in ChoiceType in 2.7.10

### ISSUE THREAD

**Issue by althaus** [symfony/symfony#18518]
Hey,

with [this commit](https://github.com/symfony/symfony/commit/0efbc30304113b987aced8f11e2d102b816a740c) @Tobion added extended handliing for placeholder in choice types. This sadly introduced a cyclic dependency with our custom EntityTypeExtension which controls our select2 binding:

```php
    public function configureOptions(OptionsResolver $resolver) {
        $resolver->setDefaults(array(
            'select2' => true,
        ));

        $resolver->setNormalizer('attr', function (Options $options, $value) {
            // Use placeholder aka empty option as placeholder for select2
            if (!empty($options['select2']) && isset($options['placeholder']) && !isset($options['data-placeholder'])) {
                $value['data-placeholder'] = $options['placeholder'];
            }

            return $value;
        });
    }
```

We basically convert a given `placeholder` to an HTML `data-placeholder` attribute. This worked in 2.7.9  and crashed since updating to 2.7.10.

```
The options "attr", "placeholder" have a cyclic dependency.
500 Internal Server Error - OptionDefinitionException

Symfony\Component\OptionsResolver\Exception\OptionDefinitionException: The options "attr", "placeholder" have a cyclic dependency.
    at n/a
        in /usr/share/iserv/vendor/symfony/symfony/src/Symfony/Component/OptionsResolver/OptionsResolver.php line 846

    at Symfony\Component\OptionsResolver\OptionsResolver->offsetGet('attr')
        in /usr/share/iserv/vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php line 356

    at Symfony\Component\Form\Extension\Core\Type\ChoiceType->Symfony\Component\Form\Extension\Core\Type\{closure}(object(OptionsResolver), null)
        in /usr/share/iserv/vendor/symfony/symfony/src/Symfony/Component/OptionsResolver/OptionsResolver.php line 962846 
etc.
```

Now I'm updating `attr` based on `placeholder` and `placeholder` wants to read `size` from `attr`. [This line](https://github.com/symfony/symfony/commit/0efbc30304113b987aced8f11e2d102b816a740c#diff-11ee91a32c601e8e1d509117556d53b3R346) must be the reason for the error.

Cheers
Matthias

**Comment by Tobion** [comment]
Hm I don't see what symfony can change about that. There is always a possibility for cyclic dependencies   as custom code can add any dependency to form options.

**Comment by Tobion** [comment]
With #4833 this problem could probably be avoided as it could distinguish dependencies on sub-options (attr.size vs attr.data-placeholder).

**Comment by HeahDude** [comment]
@althaus You can still move this resolution in `finishView()` so options are resolved during the test. It's actually very fragile to normalize `attr` for the reason given by Tobias.
What do you think ?

**Comment by althaus** [comment]
@Tobion Yes, I undestand your point of view. I'm pretty sure that this is the first time, that the `OptionsResolver` blew up on my code. I don't think that there are lot of normalizers on sub-options yet? I'd guess that this could cause 

@HeahDude Yes, I already implemented a workaroud as we've got some further code to handle select2 in `buildView` of our base `ChoiceTypeExtension`. I moved the `attr` resolving into that block.

I don't know if/where the options resolving for form types is documented. Adding a hint on this issue could prevent users from stumbling upon this.
